internal/impl/nats/output_kv.go

```diff
@@ -0,0 +1,223 @@
+
+type kvOutput struct {
+	urls   string
+	bucket string
+	key    *service.InterpolatedString
+	keyRaw string
+
+	authConf auth.Config
+	tlsConf  *tls.Config
+
+	log *service.Logger
+	fs  *service.FS
+
+	connMut  sync.Mutex
+	natsConn *nats.Conn
+	keyValue nats.KeyValue
+
+	shutSig *shutdown.Signaller
+}
+
+func newKVOutput(conf *service.ParsedConfig, mgr *service.Resources) (*kvOutput, error) {
+	kv := kvOutput{
+		log:     mgr.Logger(),
+		fs:      mgr.FS(),
+		shutSig: shutdown.NewSignaller(),
+	}
+
+	urlList, err := conf.FieldStringList("urls")
+	if err != nil {
+		return nil, err
+	}
+	kv.urls = strings.Join(urlList, ",")
+
+	if kv.bucket, err = conf.FieldString("bucket"); err != nil {
+		return nil, err
+	}
+
+	if kv.keyRaw, err = conf.FieldString("key"); err != nil {
+		return nil, err
+	}
+
+	if kv.key, err = conf.FieldInterpolatedString("key"); err != nil {
+		return nil, err
+	}
+
+	tlsConf, tlsEnabled, err := conf.FieldTLSToggled("tls")
+	if err != nil {
+		return nil, err
+	}
+	if tlsEnabled {
+		kv.tlsConf = tlsConf
+	}
+
+	if kv.authConf, err = AuthFromParsedConfig(conf.Namespace("auth")); err != nil {
+		return nil, err
+	}
+	return &kv, nil
+}
+
+//------------------------------------------------------------------------------
+
+func (kv *kvOutput) Connect(ctx context.Context) error {
+	kv.connMut.Lock()
+	defer kv.connMut.Unlock()
+
+	if kv.natsConn != nil {
+		return nil
+	}
+
+	var natsConn *nats.Conn
+	var err error
+
+	defer func() {
+		if err != nil && natsConn != nil {
+			natsConn.Close()
+		}
+	}()
+
+	var opts []nats.Option
+	if kv.tlsConf != nil {
+		opts = append(opts, nats.Secure(kv.tlsConf))
+	}
+	opts = append(opts, authConfToOptions(kv.authConf, kv.fs)...)
+	if natsConn, err = nats.Connect(kv.urls, opts...); err != nil {
+		return err
+	}
+
+	jsc, err := natsConn.JetStream()
+	if err != nil {
+		return err
+	}
+
+	kv.keyValue, err = jsc.KeyValue(kv.bucket)
+	if err != nil {
+		return err
+	}
+
+	kv.log.Infof("Setting values on NATS KV bucket: %s and key: %s", kv.bucket, kv.keyRaw)
+
+	kv.natsConn = natsConn
+	return nil
+}
+
+func (kv *kvOutput) disconnect() {
+	kv.connMut.Lock()
+	defer kv.connMut.Unlock()
+
+	if kv.natsConn != nil {
+		kv.natsConn.Close()
+		kv.natsConn = nil
+	}
+	kv.keyValue = nil
+}
+
+//------------------------------------------------------------------------------
+
+func (kv *kvOutput) Write(ctx context.Context, msg *service.Message) error {
+	kv.connMut.Lock()
+	keyValue := kv.keyValue
+	kv.connMut.Unlock()
+	if keyValue == nil {
+		return service.ErrNotConnected
+	}
+
+	value, err := msg.AsBytes()
+	if err != nil {
+		return err
+	}
+
+	key, err := kv.key.TryString(msg)
+	if err != nil {
+		return err
+	}
+
+	rev, err := keyValue.Put(key, value)
+	if err != nil {
+		return err
+	}
+
+	kv.log.With(
+		metaKVBucket, keyValue.Bucket(),
+		metaKVKey, key,
+		metaKVRevision, rev,
+	).Debug("Updated kv bucket entry")
```

Comment:
A minor one but I see in other log entries "kv" is uppercase. Perhaps it should be capitalised here too for consistency.